src/WebJobs.Script/Description/FunctionGenerator.cs

```diff
@@ -176,12 +206,26 @@ public static Type Generate(string functionAssemblyName, string typeName, Collec
                     il.Emit(OpCodes.Ldloc, taskLocal);
                 }
 
+                // Need to coerce.
+                if (innerReturnType != null)
+                {
+                    var m = typeof(FunctionGenerator).GetMethod("Coerce", BindingFlags.Static | BindingFlags.NonPublic | BindingFlags.Public);
+                    var m2 = m.MakeGenericMethod(innerReturnType);
+                    il.Emit(OpCodes.Call, m2);
+                }
+
                 il.Emit(OpCodes.Ret);
             }
 
             Type t = tb.CreateType();
 
             return t;
         }
+
+        public static async Task<T> Coerce<T>(Task<object> src)
+        {
+            var val = await src;
```

Comment:
Why are you awaiting the source task here? I think we should just be passing the source Task out to caller w/o awaiting.

Reply:
The problem is that Task<T> is not covariant. We can't cast Task<Derived> to Task<Base>. 


---
In reply to: [138178857](https://github.com/Azure/azure-webjobs-sdk-script/pull/1884#discussion_r138178857) [](ancestors = 138178857)